### ISSUE TITLE
[wpr-go] HTTPS requests handled incorrectly when macOS proxy settings is used

### ISSUE THREAD

**Issue by mhahmadi** [catapult-project/catapult#4167]
When macOS proxy settings (or an application such as Proxifier) is used to forward the traffic from Chrome in iOS simulator to the wpr-go server, HTTPs requests are not handled properly. The following message is printed:

2018/01/16 15:48:01 http: TLS handshake error from 127.0.0.1:58070: tls: oversized record received with length 20037

**Comment by xunjieli** [comment]
@nedn  Ned do you know if WPR used to work with MacOS + iOS simulator?

**Comment by nedn** [comment]
No, it has never worked on MacOS +iOS

**Comment by xunjieli** [comment]
When there is an proxy (in this case Proxifier) between Chrome and WprGo, Chrome will issue CONNECT to the proxy to set up a secure tunnel between Chrome and WprGo. 
WprGo doesn't handle CONNECT/HTTPS Tunneling, so you'll observe requests failing with ERR_TUNNEL_CONNECTION_FAILED in chrome://net-internals.

If you really want this feature, you can try to make WprGo's server implementation (catapult/web_page_replay_go/src/webpagereplay/proxy.go) handle CONNECT and somehow hijack & record the response. I am not sure if that's feasible.

@tombergan

**Comment by ncalexan** [comment]
Hi folks, I came to file an issue asking for exactly the behaviour @xunjieli suggested: in https://github.com/catapult-project/catapult/issues/4167#issuecomment-358386309.

I want to make it easier to run WprGo against Firefox and other Gecko-based products.  The Gecko engine doesn't have an [equivalent of Chrome's `--host-resolver-rules`](https://bugzilla.mozilla.org/show_bug.cgi?id=1523367), so it's not easy for me to port-map 80->8080 and 443->8081 inside of the product.  But it is easy for me to configure a proxy to do the port-mapping for me -- with, say, [goproxy](https://github.com/elazarl/goproxy) it's just a few lines of code to do exactly the connection hijacking suggested.  That is, I've made a trivial proxy that serves on localhost:4040 and just port-maps localhost:80->localhost:8080 and localhost:433->localhost:8081.  With a few layers of certificates, it appears to work just fine.  This will save me requiring privileges to bind 80 and 443 for WprGo.

One downside is that I need to coordinate two proxies, and in particular two certificates.  Clearly I'll be able to re-use one of the certificates with a little work.

So:
- could WprGo grow `CONNECT` support on the existing proxy at 8080?
- or could WprGo grow a separate `CONNECT`-supporting proxy on another port?
- if yes, would WprGo take a dependency on [goproxy](https://github.com/elazarl/goproxy), or would a reimplementation to reduce dependencies be required?

Thanks for your time, for making WprGo great (and really readable -- I don't speak Go and I can modify it), and for helping Firefox/Gecko out!
